stan/math/rev/functor/coupled_ode_system.hpp

```diff
@@ -400,27 +83,48 @@ struct coupled_ode_system<F, var, var> {
    *
    * @param[in] f the base ODE system functor
    * @param[in] y0 the initial state of the base ode
-   * @param[in] theta parameters of the base ode
-   * @param[in] x real data
-   * @param[in] x_int integer data
    * @param[in, out] msgs stream for messages
+   * @param[in] args other additional arguments
    */
-  coupled_ode_system(const F& f, const std::vector<var>& y0,
-                     const std::vector<var>& theta,
-                     const std::vector<double>& x,
-                     const std::vector<int>& x_int, std::ostream* msgs)
+  coupled_ode_system_impl(const F& f,
+                          const Eigen::Matrix<T_initial, Eigen::Dynamic, 1>& y0,
+                          std::ostream* msgs, const Args&... args)
       : f_(f),
         y0_(y0),
-        theta_(theta),
-        x_(x),
-        x_int_(x_int),
+        args_tuple_(args...),
+        local_args_tuple_(deep_copy_vars(args)...),
+        y0_vars_(count_vars(y0_)),
+        args_vars_(count_vars(args...)),
         N_(y0.size()),
-        M_(theta.size()),
-        size_(N_ + N_ * (N_ + M_)),
-        msgs_(msgs) {
-    for (const var& p : theta) {
-      theta_nochain_.emplace_back(new vari(p.val(), false));
-    }
+        args_adjoints_(args_vars_),
+        y_adjoints_(N_),
+        msgs_(msgs) {}
+
+  inline void zero_local_adjoints() {}
+
+  template <typename T, typename... Pargs>
+  inline void zero_local_adjoints(T& x, Pargs&... args) {
+    zero_local_adjoints(args...);
+  }
+
+  template <typename... Pargs>
+  inline void zero_local_adjoints(var& x, Pargs&... args) {
+    x.vi_->set_zero_adjoint();
+    zero_local_adjoints(args...);
+  }
+
+  template <typename... Pargs>
+  inline void zero_local_adjoints(std::vector<var>& x, Pargs&... args) {
+    for (size_t i = 0; i < x.size(); ++i)
+      x[i].vi_->set_zero_adjoint();
+    zero_local_adjoints(args...);
+  }
+
+  template <int R, int C, typename... Pargs>
+  inline void zero_local_adjoints(Eigen::Matrix<var, R, C>& x, Pargs&... args) {
```

Comment:
If I read this right, then this does not support at the moment nested data structures - so vectors of vectors and vectors of matrices are not allowed.

If that can be expanded it would be great - it's not a limitation in the sense that it wasn't possible before either.

Reply:
Yeah, this is what I realized I should go ahead and fix. We support anything in our other variadic arguments; we should support anything here.